File: Bronze/2017/January/01_dont_be_last/dont_be_last2.cpp
#include <iostream>
#include <fstream>
#include <unordered_map>
#include <climits>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    freopen("notlast.in", "r", stdin);
    freopen("notlast.out", "w", stdout);
    // map 存储每头奶牛的奶量，初始值 0 必须要给，因为有些奶牛可能不出现在产奶日志上
    unordered_map<string, int> cow_to_milk;
    cow_to_milk["Bessie"] = 0;
    cow_to_milk["Elsie"] = 0;
    cow_to_milk["Daisy"] = 0;
    cow_to_milk["Gertie"] = 0;
    cow_to_milk["Annabelle"] = 0;
    cow_to_milk["Maggie"] = 0;
    cow_to_milk["Henrietta"] = 0;

    int n;
    cin >> n;
    for (int i = 0; i < n; i++) {
        string cow;
        int milk;
        cin >> cow >> milk;
        cow_to_milk[cow] += milk;
    }


    vector<pair<string, int>> cowsList;
    for (auto& entry : cow_to_milk) {
        cowsList.push_back(entry);
    }
    // 根据奶量排序，如果第二小的奶量只有一个，那么就是答案，否则就是 Tie
    sort(cowsList.begin(), cowsList.end(), [](pair<string, int> a, pair<string, int> b) {
        return a.second < b.second;
    });

    int i = 0;
    string result = "Tie";

    while (i < cowsList.size()) {
        if (cowsList[i].second != cowsList[0].second) {
            if (i == cowsList.size() - 1 || cowsList[i].second != cowsList[i + 1].second) {
                result = cowsList[i].first;
            }
            break;
        }
        i++;
    }

    cout << result << endl;
    return 0;
}

File: Bronze/2017/January/01_dont_be_last/DontBeLast2.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

/**
 * @author Red
 * @version 1.0
 */
public class DontBeLast2 {
    static Map<String, Integer> cowToMilk = new HashMap<String, Integer>();
    static String ansCow = "Tie";

    public static void main(String[] args) throws Exception {
        // map 存储每头奶牛的奶量，初始值 0 必须要给，因为有些奶牛可能不出现在产奶日志上
        cowToMilk.put("Bessie", 0);
        cowToMilk.put("Elsie", 0);
        cowToMilk.put("Daisy", 0);
        cowToMilk.put("Gertie", 0);
        cowToMilk.put("Annabelle", 0);
        cowToMilk.put("Maggie", 0);
        cowToMilk.put("Henrietta", 0);
        BufferedReader r = new BufferedReader(new FileReader("notlast.in"));
        PrintWriter pw = new PrintWriter(new FileWriter("notlast.out"));
        int n = Integer.parseInt(r.readLine());
        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(r.readLine());
            String cow = st.nextToken();
            int milk = Integer.parseInt(st.nextToken());
            cowToMilk.put(cow, cowToMilk.get(cow) + milk);
        }
        // 根据奶量去重排序，如果第二小的奶量只有一个，那么就是答案，否则就是 Tie
        int[] sortedValues = cowToMilk.values().stream().distinct().sorted().mapToInt(Integer::intValue).toArray();
        if (sortedValues.length > 1) {
            int secondMin = sortedValues[1];
            for (Map.Entry<String, Integer> entry : cowToMilk.entrySet()) {
                if (entry.getValue() == secondMin) {
                    // 如果 ansCow ！= "Tie"，说明有两头奶牛的奶量都是第二小的
                    if ("Tie".equals(ansCow)) {
                        ansCow = entry.getKey();
                    } else {
                        ansCow = "Tie";
                        break;
                    }
                }
            }
        }
        pw.println(ansCow);
        pw.close();
    }
}
